Classes/Core/ConsoleApplication.php

```diff
@@ -0,0 +1,252 @@
+use TYPO3\CMS\Core\Package\PackageManager;
+use TYPO3\CMS\Core\Utility\ExtensionManagementUtility;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Extbase\Mvc\Cli\CommandManager;
+use TYPO3\CMS\Extbase\Mvc\Cli\Response;
+
+/**
+ * @internal
+ */
+class ConsoleApplication implements ApplicationInterface
+{
+    /**
+     * @var ConsoleBootstrap
+     */
+    private $bootstrap;
+
+    public function __construct(\Composer\Autoload\ClassLoader $classLoader)
+    {
+        $this->ensureRequiredEnvironment();
+        $this->bootstrap = ConsoleBootstrap::getInstance();
+        $this->bootstrap->initializeClassLoader($classLoader);
+    }
+
+    /**
+     * Bootstraps the minimal infrastructure, registers a request handler and
+     * then passes control over to that request handler.
+     *
+     * @param callable|null $execute
+     */
+    public function run(callable $execute = null)
+    {
+        $this->boot();
+
+        $this->bootstrap->registerRequestHandlerImplementation(RequestHandler::class);
+        $this->bootstrap->handleRequest(new ArgvInput());
+
+        if ($execute !== null) {
+            call_user_func($execute);
+        }
+
+        $this->shutdown();
+    }
+
+    /**
+     * Checks PHP sapi type and sets required PHP options
+     */
+    private function ensureRequiredEnvironment()
+    {
+        if (PHP_SAPI !== 'cli' || !isset($_SERVER['argc'], $_SERVER['argv'])) {
+            echo 'The command line must be executed with a cli PHP binary! The current PHP sapi type is "' . PHP_SAPI . '".' . PHP_EOL;
+            exit(1);
+        }
+        if (ini_get('memory_limit') !== '-1') {
+            @ini_set('memory_limit', '-1');
+        }
+        if (ini_get('max_execution_time') !== '0') {
+            @ini_set('max_execution_time', '0');
+        }
+    }
+
+    /**
+     * Bootstraps the minimal infrastructure, but does not execute any command
+     */
+    private function boot()
+    {
+        $this->defineBaseConstants();
+        $this->bootstrap->baseSetup();
+        // I want to see deprecation messages
+        error_reporting(E_ALL & ~(E_STRICT | E_NOTICE));
+
+        $this->requireLibraries();
+        $this->initializePackageManagement();
+
+        if (!class_exists(RunLevel::class)) {
+            echo sprintf('Could not initialize TYPO3 Console for TYPO3 in path %s.', PATH_site) . chr(10);
+            echo 'This most likely happened because you have a console code checkout in typo3conf/ext/typo3_console,' . chr(10);
+            echo 'but TYPO3 Console is not set up as extension. If you want to use it as extension,' . chr(10);
+            echo 'please download it from https://typo3.org/extensions/repository/view/typo3_console' . chr(10);
+            echo 'or install it properly using Composer.' . chr(10);
+            exit(1);
+        }
+
+        $this->bootstrap->setEarlyInstance(\Helhum\Typo3Console\Core\Booting\RunLevel::class, new RunLevel());
```

Comment:
The first argument can be shortened since the class is already imported.

Reply:
good catch, thanks